site/_posts/2018-11-19-r-spark-improvements.md

```diff
@@ -0,0 +1,197 @@
+```
+
+# Copying
+
+Currently, copying data to Spark using `sparklyr` is performed by persising
+data on-disk from R and reading it back from Spark. This was meant to be
+used for small datasets since there are better tools to transfer data
+into distributed storage systems. Nevertheless, many users have request
+support to transfer more data at fast speeds into Spark.
+
+Using `arrow` with `sparklyr`, we can transfer data directly from R to
+Spark without having to serialize this data in R or persist in disk.
+
+The following example copies 1M rows from R into Spark using `sparklyr`
+with and without `arrow`, there is close to a 10x improvement using `arrow`.
+
+This benchmark uses the [microbenchmark][8] R package, which runs code
+multiple times, provides stats on total execution time and plots each
+excecution time to understand the distribution over each iteration.
+
+```r
+microbenchmark::microbenchmark(
+  setup = library(arrow),
+  arrow_on = {
+    sparklyr_df <<- copy_to(sc, data, overwrite = T)
+    count(sparklyr_df) %>% collect()
+  },
+  arrow_off = {
+    if ("arrow" %in% .packages()) detach("package:arrow")
+    sparklyr_df <<- copy_to(sc, data, overwrite = T)
+    count(sparklyr_df) %>% collect()
+  },
+  times = 10
+) %T>% print() %>% ggplot2::autoplot()
+```
+```
+Unit: milliseconds
+      expr       min        lq      mean    median        uq      max neval
+  arrow_on  396.9805  427.5132  455.8761  447.8131  486.2177  523.980    10
+ arrow_off 2816.4576 3045.4154 3177.0776 3124.9995 3289.3395 3620.697    10
+```
+
+<div align="center">
+<img src="{{ site.base-url }}/img/arrow-r-spark-copying.png"
+     alt="Copying data with R into Spark with and without Arrow"
+     width="60%" class="img-responsive">
+</div>
+
+# Collecting
+
+Similarly, `arrow` with `sparklyr` can now avoid deserializing data in R
+while collecting data from Spark into R. These improvements are not as
+significant as copying data since, `sparklyr` already collects data in
+columnar format.
+
+The following benchmark collects 1M rows from Spark into R and shows that
+`arrow` can bring 2x improvements. 
+
+```r
+microbenchmark::microbenchmark(
+  setup = library(arrow),
+  arrow_on = {
+    collect(sparklyr_df)
+  },
+  arrow_off = {
+    if ("arrow" %in% .packages()) detach("package:arrow")
+    collect(sparklyr_df)
+  },
+  times = 10
+) %T>% print() %>% ggplot2::autoplot()
+```
+```
+Unit: milliseconds
+      expr      min       lq     mean   median       uq      max neval
+  arrow_on 260.5992 287.5217 298.4174 293.6197 312.2250 341.7968    10
+ arrow_off 363.2187 380.6286 426.6134 424.8232 440.3115 517.2623    10
+ ```
+
+<div align="center">
+<img src="{{ site.base-url }}/img/arrow-r-spark-collecting.png"
+     alt="Collecting data with R from Spark with and without Arrow"
+     width="60%" class="img-responsive">
+</div>
+
+# Transforming
+
+Today, custom transformations of data using R functions are performed in 
```

Comment:
Very good explanation for someone with no knowledge on either R internals and spark third party language support.